pkg/controller/sls/project_controller.go

```diff
@@ -0,0 +1,130 @@
+	"github.com/crossplane/crossplane-runtime/pkg/logging"
+	"github.com/crossplane/crossplane-runtime/pkg/reconciler/managed"
+	"github.com/crossplane/crossplane-runtime/pkg/resource"
+	"github.com/pkg/errors"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	slsv1alpha1 "github.com/crossplane/provider-alibaba/apis/sls/v1alpha1"
+	"github.com/crossplane/provider-alibaba/apis/v1alpha1"
+	slsclient "github.com/crossplane/provider-alibaba/pkg/clients/sls"
+)
+
+const (
+	errNotSLSProject            = "managed resource is not an SLS project custom resource"
+	errNoProvider               = "no provider config or provider specified"
+	errGetProviderConfig        = "cannot get provider config"
+	errTrackUsage               = "cannot track provider config usage"
+	errNoConnectionSecret       = "no connection secret specified"
+	errGetConnectionSecret      = "cannot get connection secret"
+	errFmtUnsupportedCredSource = "credentials source %q is not currently supported"
+)
+
+// SetupSLSProject adds a controller that reconciles SLSProjects.
+func SetupSLSProject(mgr ctrl.Manager, l logging.Logger) error {
+	options := []managed.ReconcilerOption{
+		managed.WithExternalConnecter(&connector{
+			client:      mgr.GetClient(),
+			usage:       resource.NewProviderConfigUsageTracker(mgr.GetClient(), &v1alpha1.ProviderConfigUsage{}),
+			NewClientFn: slsclient.NewClient,
+		})}
+
+	return BaseSetupSLSProject(mgr, l, options...)
+}
+
+type connector struct {
+	client      client.Client
+	usage       resource.Tracker
+	NewClientFn func(accessKeyID, accessKeySecret, region string) *slsclient.LogClient
+}
+
+func (c *connector) Connect(ctx context.Context, mg resource.Managed) (managed.ExternalClient, error) { //nolint:gocyclo
+	cr, ok := mg.(*slsv1alpha1.SLSProject)
+	if !ok {
+		return nil, errors.New(errNotSLSProject)
+	}
+
+	// provider has more than one kind of managed resource.
+	var (
+		sel    *xpv1.SecretKeySelector
+		region string
+	)
+
+	switch {
+	case cr.GetProviderConfigReference() != nil:
+		if err := c.usage.Track(ctx, mg); err != nil {
+			return nil, errors.Wrap(err, errTrackUsage)
+		}
+
+		pc := &v1alpha1.ProviderConfig{}
+		if err := c.client.Get(ctx, types.NamespacedName{Name: cr.Spec.ProviderConfigReference.Name}, pc); err != nil {
+			return nil, errors.Wrap(err, errGetProviderConfig)
+		}
+		if s := pc.Spec.Credentials.Source; s != xpv1.CredentialsSourceSecret {
+			return nil, errors.Errorf(errFmtUnsupportedCredSource, s)
+		}
+		sel = pc.Spec.Credentials.SecretRef
+		region = pc.Spec.Region
+	default:
+		return nil, errors.New(errNoProvider)
+	}
+
+	if sel == nil {
+		return nil, errors.New(errNoConnectionSecret)
+	}
+
+	s := &corev1.Secret{}
+	nn := types.NamespacedName{Namespace: sel.Namespace, Name: sel.Name}
+	if err := c.client.Get(ctx, nn, s); err != nil {
+		return nil, errors.Wrap(err, errGetConnectionSecret)
+	}
```

Comment:
@hasheddan Recently added some functionality to help extract credentials from a `ProviderConfig`. See https://github.com/crossplane/provider-template/blob/c38f3f9/internal/controller/mytype/mytype.go#L93 as an example.

Reply:
I will create another PR to utilized the common function. It needs much work:
- bump crossplane runtime version to `v0.13.0`
- throw around `ProviderConfig`
- refactor secret data for Alibaba provider
- use the common function to extract credentials and refactor `NewClient` parameters

Reply:
Here is the PR https://github.com/crossplane/provider-alibaba/pull/64